sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -380,7 +380,7 @@ object ScalaReflection extends ScalaReflection {
     val clsName = getClassNameFromType(tpe)
     val walkedTypePath = s"""- root class: "${clsName}"""" :: Nil
     extractorFor(inputObject, tpe, walkedTypePath) match {
-      case s: CreateNamedStruct => s
+      case expressions.If(IsNull(_), expressions.Literal(null, _), s: CreateNamedStruct) => s
```

Comment:
how about checking the type?
```
case i: expressions.If if tpe <:< localTypeOf[Product] => i.falseValue.asInstanceOf[CreateNamedStruct]
```
it looks safer to me.

as option of product will also